EDITD/kubetools
# Add cron job object

**## Purpose of PR**
I added CronJob object to Kubetools.


**## Parts of the app this will impact**
With this Branch we can run the following commands:
- `kubetools deploy ...`
- `kubetools remove ...`
- `kubetools show ...`
- `kubetools cleanup ...`


**## How to test it**
1.  if you have a Mac, you can test it locally by enabling Kubernetes in Docker-Desktop ( Open Docker -> Preferences -> Kubernetes -> Check **Enable Kubernetes** -> Apply & Restart) Otherwise, by pointing to the **staging**  Kubernetes  cluster

2. To get the name of the context, run this following command: `kubectl config current-context`


3. Checkout to this branch, create a python Virtual Environment, run the following command: `pip install -e .[dev]`
_Note: Run `pip install -e ".[dev]"` if you use zsh_

4. Create a **kubetools.yml** file. Content example:

```
name: my-app
namespace: cronjob-test
cronjobs:
  my-cronjob:
    schedule: "*/1 * * * *"
    concurrency_policy: "Replace"
    containers:
      hello:
        image: busybox
        command: [/bin/sh, -c, date; echo Hello from the Kubernetes cluster]
```

5. Run this command: `kubetools --context <CONTEXT NAME> deploy cronjob-test`
This command can take 20s to finish running (it waits until the first job is created)

6. The `deploy` command should create the `cronjob-test` namespace and the `my-cronjob` CronJob, you can check by running this command: `kubetools --context <CONTEXT NAME> show cronjob-test`

7. Run this command to delete the CronJob: `kubetools --context <CONTEXT NAME> remove cronjob-test`
After running this command, if you run this following command `kubetools --context <CONTEXT NAME> show cronjob-test`, the output should be `Nothing to be found here 👀!`

8. To clean the succeeded/completed pods, you can run this following command:
`kubetools --context <CONTEXT NAME> cleanup cronjob-test`


**## Additional information**
With the current code, the `kubetools cleanup` command cleans only if the pods (created by the CronJob) are orphaned. If the CronJob still exists (Not removed), even if the pods are completed they are not cleaned.
